src/objects.c

```diff
@@ -387,6 +387,42 @@ PgUser *add_user(const char *name, const char *passwd)
 	return user;
 }
 
+/* find and use original or preconfigured user given the login user */
+void use_original_user(PgUser **user)
+{
+	/*
+	 * The original user will either be the current login user, a preconfigured user via
+	 * ini file [users] section, a preconfigured user via global auth_file, a PAM user,
+	 * a user attached to a database via auth_query, a forced user via user= entry via
+	 * ini file [databases] section, or not exist yet if the user has never connected.
+	 */
+	PgUser *login_user = *user;
+	PgUser *original_user = find_user(login_user->name);
+
+	/* add new login user to reload their configurations at runtime */
+	if (original_user == NULL) {
+		put_in_order(&login_user->head, &user_list, cmp_user);
+		aatree_insert(&user_tree, (uintptr_t)login_user->name, &login_user->tree_node);
+		return;
+	}
+
+	/* original user is no longer preconfigured and won't send auth queries on future login */
+	if (original_user->is_preconfigured)
+		original_user->is_preconfigured = false;
+
+	char *tmp_passwd = strdup(login_user->passwd);
+	safe_strcpy(original_user->passwd, tmp_passwd, sizeof(original_user->passwd));
+	free(tmp_passwd);
+
+	if (login_user != original_user) {
+		statlist_remove(&user_list, &login_user->head);
```

Comment:
We deallocate any new login user that does not match the initial user reference parsed from `parse_user`. We always keep the original reference in the global `user_list`.